### ISSUE TITLE
 Argument 1 to "join" of "str" has incompatible type "Iterable[Optional[str]]"; expected "Iterable[str]"

### ISSUE THREAD

**Issue by alexey-tereshenkov-oxb** [python/mypy#8850]
## Bug/false positive

I have read https://github.com/python/mypy/issues/3057 and understand that I have to use an `Iterable` or a `Sequence` to avoid the issue with invariant classes however I still get the error.

```python
from typing import Iterable, Optional


def get(value) -> Iterable[Optional[str]]:
    if value > 5:
        return ['a', 'b']
    return []


def do(items: Iterable[Optional[str]]):
    return ','.join(items)


items = get(6)
assert do(items) == 'a,b'

items = get(3)
assert do(items) == ''
```

## What is the actual behavior/output?

> src/bla.py:11: error: Argument 1 to "join" of "str" has incompatible type "Iterable[Optional[str]]"; expected "Iterable[str]"
> Found 1 error in 1 file (checked 1 source file)

If I change the code to remove `Optional,` `mypy` doesn't report any issues. However, I feel that it is necessary to provide `Optional` to tell that the list returned by the `get` function can be either empty or containing some strings - if I understand correctly, that's the point of `Optional.`

Is there any better, more correct way to type hint this code?

```python
from typing import Iterable


def get(value) -> Iterable[str]:
    if value > 5:
        return ['a', 'b']
    return []


def do(items: Iterable[str]):
    return ','.join(items)


items = get(6)
assert do(items) == 'a,b'

items = get(3)
assert do(items) == ''

```

## What are the versions of mypy and Python you are using?

> mypy 0.770; Python 3.7.6

## What are the mypy flags you are using? (For example --strict-optional)

No flags, simply `mypy`

**Comment by srittau** [comment]
`Optional[X]` is just an alias for `Union[None, X]`. `Iterable[Optional[str]]` is an iterable that contains strings or `None` values. An empty list is still a `List[X]` for every `X`, so removing optional is correct.
